source/extensions/filters/http/cache/cache_insert_queue.cc

```diff
@@ -0,0 +1,184 @@
+    : dispatcher_(dispatcher), insert_context_(std::move(insert_context)),
+      high_watermark_bytes_(high_watermark_bytes), low_watermark_bytes_(low_watermark_bytes),
+      over_high_watermark_callback_(high), under_low_watermark_callback_(low),
+      abort_callback_(abort) {}
+
+void CacheInsertQueue::insertHeaders(const Http::ResponseHeaderMap& response_headers,
+                                     const ResponseMetadata& metadata, bool end_stream) {
+  if (end_stream) {
+    end_in_sight_ = true;
+  }
+  // While zero isn't technically true for the size of headers, headers are
+  // typically excluded from the stream buffer limit.
+  insert_context_->insertHeaders(
+      response_headers, metadata,
+      [this, end_stream](bool ready) { onChunkComplete(ready, end_stream, 0); }, end_stream);
+  chunk_in_flight_ = true;
+}
+
+void CacheInsertQueue::insertBody(const Buffer::Instance& chunk, bool end_stream) {
+  if (end_stream) {
+    end_in_sight_ = true;
+  }
+  if (chunk_in_flight_) {
+    size_t sz = chunk.length();
+    queue_size_bytes_ += sz;
+    chunks_.push_back(std::make_unique<CacheInsertChunkBody>(chunk, end_stream));
+    if (!sent_watermark_ && queue_size_bytes_ > high_watermark_bytes_) {
+      over_high_watermark_callback_();
+      sent_watermark_ = true;
+    }
+  } else {
+    insert_context_->insertBody(
+        Buffer::OwnedImpl(chunk),
+        [this, end_stream](bool ready) { onChunkComplete(ready, end_stream, 0); }, end_stream);
+    chunk_in_flight_ = true;
+  }
+}
+
+void CacheInsertQueue::insertTrailers(const Http::ResponseTrailerMap& trailers) {
+  end_in_sight_ = true;
+  if (chunk_in_flight_) {
+    chunks_.push_back(std::make_unique<CacheInsertChunkTrailers>(trailers));
+  } else {
+    insert_context_->insertTrailers(trailers,
+                                    [this](bool ready) { onChunkComplete(ready, true, 0); });
+    chunk_in_flight_ = true;
+  }
+}
+
+void CacheInsertQueue::onChunkComplete(bool ready, bool end_stream, size_t sz) {
+  dispatcher_.post([this, ready, end_stream, sz]() {
+    chunk_in_flight_ = false;
+    if (aborting_) {
+      // Parent filter was destroyed, so we can quit this operation.
+      self_ownership_.reset();
+      return;
+    }
+    ASSERT(queue_size_bytes_ >= sz);
+    queue_size_bytes_ -= sz;
+    if (sent_watermark_ && queue_size_bytes_ <= low_watermark_bytes_) {
+      under_low_watermark_callback_();
+      sent_watermark_ = false;
+    }
+    if (!ready) {
+      // canceled by cache; unwatermark if necessary, inform the filter if
+      // it's still around, and delete the queue.
+      if (sent_watermark_) {
+        under_low_watermark_callback_();
+      }
+      chunks_.clear();
+      self_ownership_.reset();
+      abort_callback_();
+      return;
+    }
+    if (end_stream) {
+      // We should only complete with end_stream with an empty queue for obvious reasons.
+      ASSERT(chunks_.empty());
```

Comment:
We could also consider ENVOY_BUG/RELEASE_ASSERT here though it seems like if the response is we ignore the insanity and clean up that's better?

Reply:
I think ASSERT is the right thing, because it's a situation that should be impossible. We want that to crash in debug builds, and to not waste time checking impossible conditions in release builds.